src/app/reporting/Engine.cpp

```diff
@@ -0,0 +1,171 @@
+namespace reporting {
+CHIP_ERROR Engine::Init()
+{
+    mMoreChunkedMessages = false;
+    mNumReportsInFlight  = 0;
+    mCurReadHandlerIdx   = 0;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Engine::BuildAndSendSingleReportData(ReadHandler * apReadHandler)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::System::PacketBufferTLVWriter reportDataWriter;
+    ReportData::Builder reportDataBuilder;
+    chip::System::PacketBufferHandle bufHandle = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLengthBytes);
+
+    VerifyOrExit(!bufHandle.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+    reportDataWriter.Init(std::move(bufHandle));
+
+    // Create a report data.
+    err = reportDataBuilder.Init(&reportDataWriter);
+    SuccessOrExit(err);
+
+    // TODO: Fill in the EventList.
+    // err = BuildSingleReportDataEventList(reportDataBuilder, apReadHandler);
+    // SuccessOrExit(err);
+
+    // TODO: Add mechanism to set mSuppressResponse to handle status reports for multiple reports
+    // TODO: Add more chunk message support, currently mMoreChunkedMessages is always false.
+    if (mMoreChunkedMessages)
+    {
+        reportDataBuilder.MoreChunkedMessages(mMoreChunkedMessages);
+    }
+
+    reportDataBuilder.EndOfReportData();
+    SuccessOrExit(reportDataBuilder.GetError());
+
+    err = reportDataWriter.Finalize(&bufHandle);
+    SuccessOrExit(err);
+
+#if CHIP_CONFIG_IM_ENABLE_SCHEMA_CHECK
+    {
+        ChipLogDetail(DataManagement, "<RE> Dumping report data...");
+        chip::System::PacketBufferTLVReader reader;
+        ReportData::Parser report;
+
+        reader.Init(bufHandle.Retain());
+        reader.Next();
+
+        err = report.Init(reader);
+        SuccessOrExit(err);
+
+        err = report.CheckSchemaValidity();
+        SuccessOrExit(err);
+    }
+#endif // CHIP_CONFIG_IM_ENABLE_SCHEMA_CHECK
+
+    ChipLogDetail(DataManagement, "<RE> Sending report...");
+    err = SendReport(apReadHandler, std::move(bufHandle));
+    VerifyOrExit(err == CHIP_NO_ERROR, ChipLogError(DataManagement, "<RE> Error sending out report data with %d!", err));
+
+    ChipLogDetail(DataManagement, "<RE> ReportsInFlight = %u with readHandler %u, RE has %s", mNumReportsInFlight,
+                  mCurReadHandlerIdx, mMoreChunkedMessages ? "more messages" : "no more messages");
+
+    if (!mMoreChunkedMessages)
+    {
+        OnReportConfirm();
+    }
+
+exit:
+    ChipLogFunctError(err);
+    if (!mMoreChunkedMessages || err != CHIP_NO_ERROR)
+    {
+        apReadHandler->Shutdown();
+    }
+    return err;
+}
+
+void Engine::Run(System::Layer * aSystemLayer, void * apAppState, System::Error)
+{
+    Engine * const pEngine = reinterpret_cast<Engine *>(apAppState);
+    pEngine->Run();
+}
+
+void Engine::ScheduleRun()
+{
+    if (InteractionModelEngine::GetInstance()->GetExchangeManager() != nullptr)
+    {
+        InteractionModelEngine::GetInstance()->GetExchangeManager()->GetSessionMgr()->SystemLayer()->ScheduleWork(Run, this);
+    }
```

Comment:
should we log an error if we cannot schedule a run?

Reply:
yes, agree, done, thanks